src/main/java/backend/RepoIO.java

```diff
@@ -51,14 +52,32 @@ public RepoIO(boolean isTestMode, boolean enableTestJSON) {
 
     public CompletableFuture<Model> openRepository(String repoId) {
         if (repoStore.isRepoStored(repoId)) {
-            return repoStore.loadRepository(repoId)
-                .thenCompose(this::updateModel)
-                .exceptionally(withResult(new Model(repoId)));
+            return loadRepositoryFromStore(repoId);
         } else {
-            return repoSource.downloadRepository(repoId)
+            return downloadRepositoryFromSource(repoId);
+        }
+    }
+
+    private CompletableFuture<Model> loadRepositoryFromStore(String repoId) {
+        return repoStore.loadRepository(repoId)
+                .thenCompose(this::updateModel)
+                .exceptionally(ex -> {
+                    try {
```

Comment:
I was bumping into something really interesting while refactoring that highlights the side-effects of things which I think would be beneficial to everyone to know.

So I'm trying to reuse `withResult` with the new `downloadRepoFromSourceBlocking(...)` and did something like this:

```java
...
return loadRepoFromStoreAsync(repoId)
        .exceptionally(withResult(downloadRepoFromSourceBlocking(repoId)));
...
```

The program ended up always downloading the repo no matter what. The reason is that java is not lazily evaluated and `downloadRepoFromSourceBlocking(repoId)` gets evaluated immediately as an argument to `withResult` and hence downloads the repo straightaway whenever `openRepository` is called.